src/server/server.rs

```diff
@@ -123,36 +120,41 @@ impl<T: RaftStoreRouter, S: StoreAddrResolver> Server<T, S> {
             ch: ch,
             storage: storage,
             end_point_worker: end_point_worker,
-            snap_mgr: snap_mgr,
+            end_point: Some(end_point),
             snap_worker: snap_worker,
+            snap_handler: Some(snap_handler),
             raft_client: RaftClient::new(env),
         };
 
         Ok(svr)
     }
 
-    pub fn run(&mut self, event_loop: &mut EventLoop<Self>) -> Result<()> {
-        let end_point = EndPointHost::new(self.storage.get_engine(),
-                                          self.end_point_worker.scheduler(),
-                                          self.cfg.end_point_concurrency,
-                                          self.cfg.end_point_txn_concurrency_on_busy,
-                                          self.cfg.end_point_small_txn_tasks_limit);
-        box_try!(self.end_point_worker.start_batch(end_point, DEFAULT_COPROCESSOR_BATCH));
-        let ch = self.get_sendch();
-        let snap_runner = SnapHandler::new(self.env.clone(),
-                                           self.snap_mgr.clone(),
-                                           self.ch.raft_router.clone(),
-                                           ch);
-        box_try!(self.snap_worker.start(snap_runner));
+    pub fn run(&mut self) -> Result<()> {
+        box_try!(self.end_point_worker
+            .start_batch(self.end_point.take().unwrap(), DEFAULT_COPROCESSOR_BATCH));
+        box_try!(self.snap_worker.start(self.snap_handler.take().unwrap()));
         self.grpc_server.start();
         info!("TiKV is ready to serve");
 
-        try!(event_loop.run(self));
-        Ok(())
-    }
+        loop {
+            match self.msg_receiver.recv().unwrap() {
+                Msg::Quit => break,
+                Msg::SendStore { store_id, msg } => self.send_store(store_id, msg),
+                Msg::ResolveResult { store_id, sock_addr, msg } => {
+                    self.on_resolve_result(store_id, sock_addr, msg)
+                }
+                Msg::CloseConn { .. } => {}
```

Comment:
gRPC conn doesn't need to close manually, it is kept now only for compatibility.